### PR TITLE
[Sharktank] gpt-oss configs

### PR DESCRIPTION
**Pull request by oyazdanb**
cleaning llm.py and modifying llm_configs.py to prepare for gpt-oss

### CODE REVIEW

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 1]
```diff
@@ -300,7 +340,29 @@ def get_custom_configs(p: dict[str, Any], name_prefix: str):
         res["expert_shared_feed_forward_length"] = _int_prop(
             p, f"{name_prefix}.expert_shared_feed_forward_length"
         )
-
+        res["use_selective_moe"] = True
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:343]: `deepseek2` also needs `use_selective_moe`. The logic of selecting moe is different between llama4 (even/odd layers are moe) and deepseek (first `n_dense_layers` are not moe and all following blocks it are moe). Logic to calculate `is_moe_block` can be consolidated for both models on `llm.py`.

File: sharktank/sharktank/models/llm/llm.py [hunk 2]
```diff
@@ -303,45 +315,37 @@ def __init__(
...
+        # Use config flags for FFN behavior
+        self.add_residual = config.hp.use_ffn_residual
+        self.use_ffn_norm = config.hp.use_ffn_norm
 
         is_moe_block = False
-        experts_ffn_moe_block = "DenseFFNMOE"
-        if config.hp.model_arch == "llama4":
+        if config.hp.use_selective_moe:
+            # Model uses MoE selectively based on layer index
             is_moe_block = block_index in config.moe_layers
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/llm/llm.py:343]: Logic to calculate `is_moe_block` can be consolidated for both llama4 and deepseeek2 (Lines 351-352).

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 3]
```diff
@@ -120,6 +120,34 @@ class LlamaHParams:
     # Scaling factor applied as a floor value in attention computations.
     floor_scale: Optional[int] = None
 
+    # gpt-oss configs
+    use_fused_swiglu: bool = False  # Whether to use fused swiglu processing
+    sliding_window: int = 0  # 0 = no sliding window, >0 = window size
+    swiglu_limit: Optional[float] = None
+    rope_gpt_oss: bool = False
+    use_fused_qkv: bool = False
+    use_direct_expert_routing: bool = False
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:129]: Can you add a comment for as many args as possible?

File: sharktank/sharktank/layers/mixture_of_experts_block.py [hunk 4]
```diff
@@ -137,61 +150,81 @@ def forward(
         batch_size, sequence_length, feature_dim = h.shape
         ffn_input = h.view(-1, feature_dim)
 
+        # Apply input normalization for direct routing
+        if self.use_direct_expert_routing:
+            router_input = self.ffn_norm_scale(ffn_input)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/mixture_of_experts_block.py:155]: If `ffn_norm_scale` is only used when `use_direct_expert_routing` is True, then we can remove L140-143 above.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 5]
```diff
@@ -120,6 +120,34 @@ class LlamaHParams:
     # Scaling factor applied as a floor value in attention computations.
     floor_scale: Optional[int] = None
 
+    # gpt-oss configs
+    use_fused_swiglu: bool = False  # Whether to use fused swiglu processing
+    sliding_window: int = 0  # 0 = no sliding window, >0 = window size
+    swiglu_limit: Optional[float] = None
+    rope_gpt_oss: bool = False
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:127]: Can we rename this to more functional than model oriented?

File: sharktank/sharktank/layers/mixture_of_experts_block.py [hunk 6]
```diff
@@ -137,61 +150,81 @@ def forward(
...
+        moe_output = self.routed_experts(router_input, top_k_experts, expert_gate)
 
         if self.expert_shared_count is not None:
             moe_output = moe_output + self.shared_experts(ffn_input)
 
         moe_output = moe_output.reshape(batch_size, sequence_length, feature_dim)
-
-        moe_output = self.layer_output_norm(moe_output)
+        if self.use_direct_expert_routing:
+            moe_output = moe_output + h
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/mixture_of_experts_block.py:217]: added a new variable use_residual_moe to generalize it.

**Review by rsuderman — CHANGES_REQUESTED** [review]
I am worried about how much rearchitecting is in this PR. A lot of the changes needed here are routing variables / configurations. We really should try to split these into the configurations being landed rather than having a single monolothic commit. This could easily introduce regressions in other moels that are hard to detect.

File: sharktank/sharktank/layers/mixture_of_experts_block.py [hunk 7]
```diff
@@ -137,61 +145,77 @@ def forward(
...
-        moe_output = self.routed_experts(ffn_input, top_k_experts, expert_gate)
+        moe_output = self.routed_experts(router_input, top_k_experts, expert_gate)
 
         if self.expert_shared_count is not None:
             moe_output = moe_output + self.shared_experts(ffn_input)
 
         moe_output = moe_output.reshape(batch_size, sequence_length, feature_dim)
-
-        moe_output = self.layer_output_norm(moe_output)
+        if self.use_residual_moe:
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/mixture_of_experts_block.py:218]: added a new variable to take care of normalization.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 8]
```diff
@@ -120,6 +120,37 @@ class LlamaHParams:
     # Scaling factor applied as a floor value in attention computations.
     floor_scale: Optional[int] = None
 
+    # gpt-oss configs
+    use_moe_swiglu: bool = (
+        False  # Whether to use swiglu activation in MoE blocks instead of silu
+    )
+    sliding_window: int = 0  # 0 = no sliding window, >0 = window size
+    swiglu_limit: Optional[float] = None  # Limit for swiglu activation function
+    rope_gpt_oss: bool = False  # Whether to use gpt-oss RoPE: disables interleaved mode, uses base^(i/d) freqs, applies gpt-oss YaRN variant
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/configs/llm_configs.py]: gpt-oss uses interweaved for rope; i decoupled it from use_hf by adding a new parameter.

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 9]
```diff
@@ -120,6 +120,37 @@ class LlamaHParams:
...
+    use_fused_qkv: bool = False  # Whether to use fused QKV layer instead of separate Q,K,V in attention preprocessing
+    use_direct_expert_routing: bool = (
+        False  # Whether to use simplified MoE routing without expert groups
+    )
+
+    # MoE architecture configuration
+    is_moe_model: bool = False  # Whether this model uses MoE for all layers
+    moe_block_type: str = "DenseFFNMOE"  # Type of MoE block to use
```
[the remark this answers is not in the document]
Comment (oyazdanb) [reply sharktank/sharktank/layers/configs/llm_configs.py]: moe_block is defined based on model_arch now; we need to work on reading it from params in another PR.

File: sharktank/sharktank/models/llm/llm.py [hunk 10]
```diff
@@ -378,6 +402,10 @@ def __init__(
                 ),
             )
 
+        # Set FFN configuration attributes
+        self.use_ffn_norm = getattr(config.hp, "use_ffn_norm", True)
+        self.use_ffn_residual = getattr(config.hp, "use_ffn_residual", True)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/llm/llm.py:407]: Have 3 things to clarify here:
- Is `use_ffn_residual` == `add_residual`?
- for gpt-oss, `add_residual` = `True` but `use_ffn_residual` = False 
- `use_ffn_residual` isn't used anywhere

File: sharktank/sharktank/layers/configs/llm_configs.py [hunk 11]
```diff
@@ -130,6 +130,27 @@ class LlamaHParams:
     # Scaling factor applied as a floor value in attention computations.
     floor_scale: Optional[int] = None
 
+    # gpt-oss configs
+    use_moe_swiglu: bool = (
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/configs/llm_configs.py:134]: I see `swiglu` defined in `llm.py` for this model, is there a reason we need this separate config var?

File: sharktank/sharktank/layers/rotary_embedding_hf.py [hunk 12]
```diff
@@ -190,17 +186,17 @@ def _apply_yarn(self, freqs):
             freqs = torch.where(is_medium_freq, smoothed_inv_freq, inv_freq_llama)
         return freqs
 
-    def _apply_yarn_openweight(self, freqs):
+    def _apply_yarn_gpt_oss(self, freqs):
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/rotary_embedding_hf.py:189]: Is it possible to fold `_apply_yarn_openweight()` into `_apply_yarn()` or even rename it generically, without including the model name.

File: sharktank/sharktank/layers/rotary_embedding_hf.py [hunk 13]
```diff
@@ -126,14 +122,14 @@ def _compute_theta(self, device):
         #   theta = 10000^{-2 (i - 1) / d}, i \in [1, 2, ..., d/2]
         # which is a convoluted way of saying
         #   theta = (1/base)^{i / d}, i \in range(0, dim, 2)
-        if self.rope_openweight:
-            # OpenWeight base freqs:base^(i/d)
+        if self.rope_gpt_oss:
+            # gpt-oss base freqs:base^(i/d)
             freqs = self.rope_theta ** (
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/rotary_embedding_hf.py:127]: We can consolidate the yarn branching in `_compute_theta()`, eg: `dim == self.head_dim` and the freq base is same for both yarn variations.

File: sharktank/sharktank/layers/rotary_embedding.py [hunk 14]
```diff
@@ -160,14 +160,20 @@ def build_rotary_layer(
...
+    rope_interleaved: Optional[bool] = None,
     **rotary_embd_layer_kwargs,
 ) -> CachedRotaryLayer:
     rope_freq_base = 10000.0 if rope_freq_base is None else rope_freq_base
 
     rotary_embd_layer_kwargs = rotary_embd_layer_kwargs.copy()
     rotary_embd_layer_kwargs["rope_theta"] = rope_freq_base
     rotary_embd_layer_kwargs["head_dim"] = rope_dimension_count
-    rotary_embd_layer_kwargs["interleaved"] = not use_hf
+    # Use rope_interleaved if provided, otherwise fall back to not use_hf for backward compatibility
```
Comment (archana-ramalingam) [inline sharktank/sharktank/layers/rotary_embedding.py:172]: Can remove the comment as we have described the values already in configs.

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 5 above)
Comment (archana-ramalingam) [reply sharktank/sharktank/layers/configs/llm_configs.py]: In general, we want to have generic/function oriented naming rather than a model specific one, as it can be reused in future models.

File: sharktank/sharktank/layers/configs/llm_configs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (archana-ramalingam) [reply sharktank/sharktank/layers/configs/llm_configs.py]: For context, `DenseFFNMOE`, `SparseFFNMOE`, `PreGatherFFNMOE` are all functionally the same. 
`PreGatherFFNMOE` worked for grok, mixtral but for larger models like deepseek, we needed DenseFFNMOE to resolve memory issues. 
`SparseFFNMOE` was the very first version overridden by PreGatherFFNMOE for performance.

Ref: https://github.com/nod-ai/shark-ai/pull/1359

File: sharktank/sharktank/layers/rotary_embedding_hf.py (hunk 13 above)
Comment (oyazdanb) [reply sharktank/sharktank/layers/rotary_embedding_hf.py:127]: the frequencey used in two branches are different and yarn methods expects different input formats. cosolidating them makes it more complex.

**Review by rsuderman — APPROVED** [review]
